### ISSUE TITLE
How to instruct BDWGC to scan manually-allocated stacks in coroutines

### ISSUE THREAD

**Issue by nickwanninger** [ivmai/bdwgc#274]
Is there a way to add a region of memory mapped with `mmap` to the garbage collector so it searches there for references?

**Comment by ivmai** [comment]
Do you mean adding a data root? If yes, please look at GC_add_roots()

**Comment by nickwanninger** [comment]
Hmm doesnt seem to work. I'm trying to implement a stackful coroutine system in my language which requires allocating a new c stack that I then switch in and out of (treated as python generators). Though when I use libgc I get unexpected segfaults from what I assume is incorrectly freed data. I can turn off the garbage collector and use system malloc and everything works perfectly fine.

Is there something else I need to do to specifically add a stack?

**Comment by ivmai** [comment]
Similar to #150

**Comment by ivmai** [comment]
At least the problem is how to determine the stack base of the current coroutine stack.
Coroutines are not supported out-of-the-box in BDWGC currently, so some tricks should be used.
Look at STACK_NOT_SCANNED.

**Comment by ivmai** [comment]
Also look at GC_[un]register_my_thread.

**Comment by ivmai** [comment]
I think the followings should work (STACK_NOT_SCANNED and GC_add_roots are not needed).
Each coroutine should look like:
```
void coroutine_inner(struct GC_stack_base *pbase, void *arg) {
    GC_register_my_thread(pbase); // this one records the stack base of the this coroutine
    .... // your code here
    GC_do_blocking(&yield, 0); // yield() is the function what switches to other coroutine
         // GC_do_blocking flushes all registers to stack, records current stack pointer, and instruct the GC that this thread should not be stopped during stop-the-world collection phase, and then calls yield, and it "unblocks" when yield returns execution control to this coroutine.
    .... // your code here
    GC_unregister_my_thread(); // unregisters the coroutine
}

void coroutine(void *arg) {
   GC_call_with_stack_base(coroutine_inner, arg);
}
```

**Comment by nickwanninger** [comment]
This will work if I yield out of the coroutine before it is completed as well?

**Comment by ivmai** [comment]
> This will work if I yield out of the coroutine before it is completed as well?

Do you mean yield and never return back (never do GC_unregister_my_thread)? In this case, it will stay registered in the collector and its stack will be scanned on every garbage collection.

**Comment by nickwanninger** [comment]
well so I'm using longjmps to switch between stacks. When you resume a coroutine, it sets a return point then jumps into the coroutine's jumpbuf (initialized somewhere else). Then the coroutine never really "returns" it just yields (longjmps) back to that return point then allows the `resume` function to return.

I guess I could put all this logic at the top and end of the `resume` function unless {,un}register_thread takes alot of time to do its thing

**Comment by ivmai** [comment]
Yes, it is also possible to put [un]register_my_thread around setjmp and, e.g. allocate stacks using GC_malloc_uncollectable. You should call GC_set_all_interior_pointers(1) before GC_INIT()  in this case. And you should ensure that the general-purpose registers saved by longjmp are scanned by the collector in this case.

**Comment by nickwanninger** [comment]
I appear to have gotten it working by using the following code. I'm basically playing with the stack_base struct in order to force certain mem_base. I just need to figure out where to put the `deregister_thread` call. I appreciate the help! Though I do feel like it would be nice to have the feature mentioned in #150:
> My solution now was to allocate all user-space stacks with GC_malloc_uncollectable and then to check whether the current stack pointer points to somewhere in the GC heap with GC_is_heap_ptr(lo), if yes then hi = GC_base(lo) + STACK_SIZE. Thank you again for your help.
```c
    // calculate the new stack pointer
    uint64_t sp = (uint64_t)(stk);
    sp += get_stack_size();
    sp &= ~15;

    // inline assembly macro to set the stack pointer
    SET_SP(sp);

    // instruct the GC that the new stack is being used
    struct GC_stack_base stack_base;
    GC_get_stack_base(&stack_base);
    stack_base.mem_base = (void*)sp;
    // register that new stack location to the garbage collector
    GC_register_my_thread(&stack_base);
    // call the coroutine function
    func(this);
    done = true;
    // switch back to the calling context
    longjmp(return_ctx, 1);
    return value;

```

I appreciate the help!

**Comment by ivmai** [comment]
I'm thinking of adding the following function to GC:
int GC_call_swapcontext(int (*swapcontext_fn)(void *arg1,void *arg2), void *arg1, void *arg2)

So that all you need to do is:
* place coroutines context and stack is some memory scanned by GC (e.g. put to globals or main() stack or alloc with GC_malloc_uncollectable), and
* wrap each call to context switching routine, e.g.: res = GC_call_swapcontext(swapcontext, &cur_uctx, &new_ctxt);

Should fit your needs, right?

**Comment by nickwanninger** [comment]
Hmm, how would that allow me to 'yield' then resume execution of a coroutine in the same manner as a process at the kernel level? Would it switch back after the call to swap_context?

**Comment by nickwanninger** [comment]
Ah wait, never mind. That makes sense and should work, yeah

**Comment by bcardiff** [comment]
@nickwanninger I left some notes regarding how we are handling this currently in Crystal at https://github.com/ivmai/bdwgc/pull/277#issuecomment-485525695.

@ivmai As pointed in that comment there is no need on Crystal to let the GC know about every context switch as the `GC_call_swapcontext` will imply. But if the overhead is minimum I understand it might be a better API.

**Comment by ivmai** [comment]
I have cancelled work on GC_call_swapcontext in favor of the concept proposed by @bcardiff in PR #277.

@nickwanninger Could you please test the PR #277 if it works for you?

**Comment by nickwanninger** [comment]
Yeah I can look into it, I'm still having some problems with the build system on macos under cmake, but I'll see what I can do

**Comment by ivmai** [comment]
I've finally submitted the commit proposed by @bcardiff (with plenty of modifications), see 5668de711. I had to change the API slightly to make it portable.
@nickwanninger, is is OK?

**Comment by bcardiff** [comment]
Thanks @ivmai for merging the feature upstream. There are definitely some of the changes you've done that I was not able to know they were even desired. We will try to test these additions and use them in the next release of bdwgc and crystal.

**Comment by ivmai** [comment]
@bcardiff, the most notable change against the initial variant is the use of a platform-neutral thread handle (returned by GC_get_my_stackbottom) instead of pthread. The other important change is that it is explicitly said that the client should hold the GC lock when calling G_set_stackbottom.
Other changes (use of struct GC_stack_base, renaming of GC_get_stackbottom) are quite cosmetic.

**Comment by ivmai** [comment]
> We will try to test these additions and use them in the next release of bdwgc and crystal.

The feature will be included in bdwgc v8.2.0 but the release won't happen soon, I think it will be in November (maybe earlier, it depends on activity of contribution of new features).

**Comment by raviqqe** [comment]
Hi. Thanks for the great changes! I'm currently working on [my Rust binding of bdwgc](https://github.com/raviqqe/bdwgc-alloc) and trying to add coroutines support. I have a few questions about the new API.

1. Why the `GC_set_my_stackbottom` needs manual locking while `GC_get_my_stackbottom` does it automatically?
1. Because context switches and registration of stack bottoms cannot be done at the same time and atomic, do they need to be done between calls `GC_disable` and `GC_enable` as suggested in #277 by @bcardiff ?

**Comment by bcardiff** [comment]
In Crystal the `GC_get_my_stackbottom` will be used only a couple of times, mainly to retrieve the initial stackbottom of each thread (primary and new ones) in a portable way.

On the other hand, `GC_set_my_stackbottom` is expected to be called a lot of times. There are two alternatives we've used:

a. During each context switch, disable the GC and set the new stackbottom.
b. Or, before the collection occurs, iterate all the exiting coroutines and either set their stack as roots or set the stackbottom of the running thread if the coroutine is running.

What I don't follow is

> Because context switches and registration of stack bottoms cannot be done at the same time and atomic (..)

Creation of stack bottoms is close related to how you want to implement/schedule coroutine and that is outside the responsibility of the GC. There is no registration of the stack bottom, only setting the current stack bottom of a thread.

**Comment by ivmai** [comment]
> Why the GC_set_my_stackbottom needs manual locking while GC_get_my_stackbottom does it automatically?

GC_get_my_stackbottom acquires the lock because it is needed to avoid data race during GC_lookup_thread and because it is not time-critical (it is called just once per thread as pointed out by @bcardiff).

GC_set_stackbottom alters stack bottom value in the GC internal structure for the given thread, to ensure no data race (the GC is not scanning data roots right now) the lock must be held - delegating the client to acquire the GC lock provides the ability to make a context switch atomically.

**Comment by ivmai** [comment]
Disabling of GC should not be needed (GC cannot occur if the client holds the lock).
Even in case of incremental collection, it is guaranteed that the current stack and GC_push_other_roots is done in one uninterruptible manner (holding the GC lock).

**Comment by raviqqe** [comment]
@bcardiff I meant setting stack bottoms by registration of stack bottoms. Sorry for confusion!
@ivmai So, does it mean that context switch must be done inside callbacks which are passed to `GC_call_with_alloc_lock` if I don't use `GC_disable` and `GC_enable`? I guess, otherwise, threads may have conditions of SB (Stack Bottom) >> SP (Stack Pointer) and having invalid memory regions between them.

Here are some safe and unsafe examples in my understanding.

### Safe example

1. Acquire an alloc lock (or disable GC.)
1. Set SB.
1. Set SP (context switch. 2 and 3 are interchangable, I guess)
1. Release an alloc lock (or enable GC.)

### Unsafe example

1. Set SP.
1. (At this point the memory region between SP and SB might be invalid.)
1. Acquire an alloc lock (or disable GC.)
1. Set SB.
1. Release an alloc lock (or enable GC.)

**Comment by ivmai** [comment]
@raviqqe, Please avoid using GC_disable/enable in production code - this could lead to unbounded heap growth in the multi-threaded environment (because while GC is disabled, other thread may try to allocate memory causing heap growth). Use GC locking instead.

> does it mean that context switch must be done inside callbacks which are passed to GC_call_with_alloc_lock

Yes. You can use a lambda in C++.

Safe example is OK. But please note that you release the lock in the target co-routine. (Does it work correctly in your co-routines switch implementation?)

> I would also like to know if BDW GC handles the case of SP > SB well or not.

No.

**Comment by raviqqe** [comment]
> Please avoid using GC_disable/enable in production code

If we cannot use `GC_disable` and `GC_enable`, can we add API functions like `GC_lock` and `GC_unlock` while they are a kind of anti-pattern for lock management? I think we cannot use `GC_call_with_alloc_lock` for this case because we cannot switch contexts in callbacks passed to `GC_call_with_alloc_lock`.

**Comment by ivmai** [comment]
>> Does it work correctly in your co-routines switch implementation?

I mean it in regarding locks ownership - the locks are owned by real threads, not co-routines, so if you lock something in one co-routine and then switch to another co-routine, and then do unlocking - will it proceed with unlocking successfully or fail saying "this thread is not owning the lock"?

> can we add API functions like GC_lock and GC_unlock

Yes, I've added #282 about it.

**Comment by raviqqe** [comment]
The lock management should be fine in my use case because pairs of lock and unlock are done in the same threads on both resumption and suspension of coroutines although I cannot test that currently.

**Comment by bcardiff** [comment]
Hi @ivmai, I wanted to check if there is an ETA for 8.2.0 since it will allow us to drop the patched version of the gc we've been using. Thanks!

**Comment by ivmai** [comment]
Hi.
To my current understanding, 8.2.0 will be released at the end of February.
